gateway-provider-security-shiro/src/main/java/org/apache/knox/gateway/shirorealm/KnoxPamRealm.java

```diff
@@ -161,6 +149,23 @@ protected AuthenticationInfo doGetAuthenticationInfo(AuthenticationToken token)
         credentialsHash.getSalt(), getName());
   }
 
+  private synchronized UnixUser lookupUnixUser(AuthenticationToken token) {
```

Comment:
This single threads all requests that use groups then right?

I know that libpam4j used to have threading issues under the hood. It might be there are still issues there 

see https://github.com/kohsuke/libpam4j/issues/28